pgn-reader/src/buffer.rs

```diff
@@ -1,80 +1,168 @@
+    ///
+    /// Never greater than `self.end`.
+    start: usize,
+    /// The end of the valid data + 1 (not a valid index).
+    ///
+    /// Never greater than [`CAPACITY`].
+    end: usize,
 }
 
 impl Buffer {
-    pub(crate) fn with_capacity(capacity: usize) -> Buffer {
-        Buffer {
-            buf: vec![0; capacity].into_boxed_slice(),
-            pos: 0,
-            filled: 0,
+    /// Creates a new [`Buffer`] that can hold [`CAPACITY`] many elements.
+    pub(crate) fn new() -> Self {
+        Self {
+            buffer: vec![0; CAPACITY].into_boxed_slice(),
+            start: 0,
+            end: 0,
         }
     }
 
+    /// Equivalent to [`self.data_range().len()`](Self::data_range), but faster.
+    #[inline]
+    fn data_len(&self) -> usize {
+        self.end - self.start
+    }
+
+    #[inline]
+    /// Range from `self.start` to `self.end`.
+    ///
+    /// This is where [`Self::data`] lives.
+    fn data_range(&self) -> Range<usize> {
+        self.start..self.end
+    }
+
+    /// Gets the valid data in the buffer.
     #[inline]
     pub(crate) fn data(&self) -> &[u8] {
-        // SAFETY: self.pos <= self.filled <= self.buf.len()
-        unsafe { self.buf.get_unchecked(self.pos..self.filled) }
+        debug_assert!(self.start <= self.end && self.end <= CAPACITY);
+
+        // SAFETY: self.start <= self.end <= CAPACITY
+        unsafe { self.buffer.get_unchecked(self.data_range()) }
+    }
+
+    /// Returns the first item in [`Self::data`].
+    #[inline]
+    pub(crate) fn peek(&self) -> Option<u8> {
+        self.data().first().copied()
     }
 
+    /// Clears the buffer.
     #[inline]
-    pub(crate) fn discard_data(&mut self) {
-        self.pos = 0;
-        self.filled = 0;
+    pub(crate) fn clear(&mut self) {
+        self.start = 0;
+        self.end = 0;
     }
 
+    /// Discards `n` many bytes at the front of [`Self::data`].
     #[inline]
     pub(crate) fn consume(&mut self, n: usize) {
-        self.pos = cmp::min(self.pos + n, self.filled);
+        self.start = cmp::min(self.start + n, self.end);
     }
 
+    /// Like [`self.consume(1)`](Self::consume).
     #[inline]
     pub(crate) fn bump(&mut self) {
         self.consume(1);
     }
 
-    #[inline]
-    pub(crate) fn peek(&self) -> Option<u8> {
-        self.data().first().copied()
-    }
+    /// Ensures that `N` amount of bytes are in the buffer and returns the data.
+    ///
+    /// The only situation where the returned slice does not have `N` elements is if EOF was
+    /// encountered.
+    pub(crate) fn ensure_bytes<const N: usize>(&mut self, mut r: impl Read) -> io::Result<&[u8]> {
+        const {
+            debug_assert!(N <= CAPACITY);
+        }
 
-    pub(crate) fn ensure_bytes(&mut self, n: usize, mut reader: impl Read) -> io::Result<&[u8]> {
-        debug_assert!(n < self.buf.len());
+        // more backshifts but less syscalls - net performance gain
+        if self.start > 0 {
+            self.backshift();
```

Comment:
now we backshift way too often. it's really only a useful operation if we're also going to use the new space for a read